domain/secret/service/access.go

```diff
@@ -168,33 +175,47 @@ func grantParams(in SecretAccessParams) domainsecret.GrantParams {
 // RevokeSecretAccess revokes access to the secret for the specified subject.
 // It returns an error satisfying [secreterrors.SecretNotFound] if the secret is not found.
 func (s *SecretService) RevokeSecretAccess(ctx context.Context, uri *secrets.URI, params SecretAccessParams) error {
-	if err := s.canManage(ctx, uri, params.Accessor, params.LeaderToken); err != nil {
+	isLeader, err := checkLeaderToken(params.LeaderToken, params.Accessor)
+	if err != nil {
 		return errors.Trace(err)
 	}
 
-	p := domainsecret.AccessParams{
-		SubjectID: params.Subject.ID,
+	return s.secretState.RunAtomic(ctx, func(ctx domain.AtomicContext) error {
+		if err := s.canManage(ctx, isLeader, uri, params.Accessor); err != nil {
+			return errors.Trace(err)
+		}
+
+		p := domainsecret.AccessParams{
+			SubjectID: params.Subject.ID,
+		}
+		switch params.Subject.Kind {
+		case UnitAccessor:
+			p.SubjectTypeID = domainsecret.SubjectUnit
+		case ApplicationAccessor:
+			p.SubjectTypeID = domainsecret.SubjectApplication
+		case RemoteApplicationAccessor:
+			p.SubjectTypeID = domainsecret.SubjectRemoteApplication
+		case ModelAccessor:
+			p.SubjectTypeID = domainsecret.SubjectModel
+		}
+		return s.secretState.RevokeAccess(ctx, uri, p)
+	})
+}
+
+func checkLeaderToken(token leadership.Token, assessor SecretAccessor) (bool, error) {
+	if assessor.Kind != UnitAccessor {
```

Comment:
We don't care about the accessor (param name has a typo btw) here. We just care whether the token says true or not. The logic for canManage() which uses the result (true/false) has the check for being a unit. This effectively does the check twice.